cmd/service-controller/console_server.go

```diff
@@ -0,0 +1,501 @@
+package main
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/skupperproject/skupper/client"
+	"github.com/skupperproject/skupper/pkg/qdr"
+)
+
+type ConsoleServer struct {
+	agentPool *qdr.AgentPool
+	iplookup  *IpLookup
+}
+
+func newConsoleServer(cli *client.VanClient, config *tls.Config) *ConsoleServer {
+	return &ConsoleServer{
+		agentPool: qdr.NewAgentPool("amqps://skupper-messaging:5671", config),
+		iplookup:  NewIpLookup(cli),
+	}
+}
+
+func authenticate(dir string, user string, password string) bool {
+	filename := path.Join(dir, user)
+	file, err := os.Open(filename)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Printf("Failed to authenticate %s, no such user exists", user)
+		} else {
+			log.Printf("Failed to authenticate %s: %s", user, err)
+		}
+		return false
+	}
+	defer file.Close()
+
+	bytes, err := ioutil.ReadAll(file)
+	if err != nil {
+		log.Printf("Failed to authenticate %s: %s", user, err)
+		return false
+	}
+	return string(bytes) == password
+}
+
+func authenticated(h http.Handler) http.Handler {
+	dir := os.Getenv("METRICS_USERS")
+	if dir != "" {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			user, password, _ := r.BasicAuth()
+
+			if authenticate(dir, user, password) {
+				h.ServeHTTP(w, r)
+			} else {
+				w.Header().Set("WWW-Authenticate", "Basic realm=skupper")
+				http.Error(w, "Unauthorized", http.StatusUnauthorized)
+			}
+		})
+	} else {
+		return h
+	}
+}
+
+func (server *ConsoleServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	agent, err := server.agentPool.Get()
+	if err != nil {
+		log.Printf("Could not get management agent : %s", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+	data, err := getConsoleData(agent, server.iplookup)
+	server.agentPool.Put(agent)
+	if err != nil {
+		log.Printf("Error retrieving console data: %s", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	} else {
+		bytes, err := json.MarshalIndent(data, "", "    ")
+		if err != nil {
+			log.Printf("Error writing json: %s", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		} else {
+			fmt.Fprintf(w, string(bytes))
+		}
+	}
+}
+
+func (server *ConsoleServer) start(stopCh <-chan struct{}) error {
+	err := server.iplookup.start(stopCh)
+	go server.listen()
+	return err
+}
+
+func (server *ConsoleServer) listen() {
+	http.Handle("/DATA", authenticated(server))
+	http.Handle("/", authenticated(http.FileServer(http.Dir("/app/console/"))))
+	log.Fatal(http.ListenAndServe(":8080", nil))
```

Comment:
Need to be able to setup different listen port with authmode is openshift (e.g. for oauth reverse proxy setup)...